exporter/datadogexporter/internal/logs/translator.go

```diff
@@ -0,0 +1,172 @@
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/otlp/model/attributes"
+	"github.com/DataDog/datadog-agent/pkg/otlp/model/source"
+	"github.com/DataDog/datadog-api-client-go/v2/api/datadog"
+	"github.com/DataDog/datadog-api-client-go/v2/api/datadogV2"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/plog"
+	conventions "go.opentelemetry.io/collector/semconv/v1.6.1"
+)
+
+const (
+	namespace          = "otel"
+	otelTraceID        = namespace + ".trace_id"
+	otelSpandID        = namespace + ".span_id"
+	otelSeverityNumber = namespace + ".serverity_number"
+	otelSeverityText   = namespace + ".serverity_text"
+	otelTimestamp      = namespace + ".timestamp"
+
+	ddStatus    = "status"
+	ddTraceID   = "dd.trace_id"
+	ddSpanID    = "dd.span_id"
+	ddTimestamp = "@timestamp"
+)
+
+const (
+	logLevelTrace = "trace"
+	logLevelDebug = "debug"
+	logLevelInfo  = "info"
+	logLevelWarn  = "warn"
+	logLevelError = "error"
+	logLevelFatal = "fatal"
+)
+
+// Transform is responsible to convert LogRecord to datadog format
+func Transform(lr plog.LogRecord, res pcommon.Resource) datadogV2.HTTPLogItem {
+	hostName, serviceName := extractHostNameService(res.Attributes(), lr.Attributes())
+
+	l := datadogV2.HTTPLogItem{
+		AdditionalProperties: make(map[string]string),
+	}
+	if hostName != "" {
+		l.Hostname = datadog.PtrString(hostName)
+	}
+
+	if serviceName != "" {
+		l.Service = datadog.PtrString(serviceName)
+	}
+
+	// we need to set log attributes as AdditionalProperties
+	lr.Attributes().Range(func(k string, v pcommon.Value) bool {
+		l.AdditionalProperties[k] = v.AsString()
+		return true
+	})
+	if !lr.TraceID().IsEmpty() {
+		l.AdditionalProperties[ddTraceID] = convertTraceID(lr.TraceID().HexString())
+		l.AdditionalProperties[otelTraceID] = lr.TraceID().HexString()
+	}
+	if !lr.SpanID().IsEmpty() {
+		l.AdditionalProperties[ddSpanID] = convertTraceID(lr.SpanID().HexString())
+		l.AdditionalProperties[otelSpandID] = lr.SpanID().HexString()
+	}
+	var status string
+
+	if lr.SeverityText() != "" {
+		status = lr.SeverityText()
+		l.AdditionalProperties[otelSeverityText] = lr.SeverityText()
+	} else {
+		status = deriveStatus(lr.SeverityNumber())
+	}
+	l.AdditionalProperties[ddStatus] = status
+
+	if lr.SeverityNumber() != 0 {
+		l.AdditionalProperties[otelSeverityNumber] = fmt.Sprintf("%d", lr.SeverityNumber())
+	}
+
+	// for datadog to use the same timestamp we need to set the additional property of "@timestamp"
+	if lr.Timestamp() != 0 {
+		// we are retaining the nano second precision in this property
+		l.AdditionalProperties[otelTimestamp] = fmt.Sprintf("%d", lr.Timestamp())
+		l.AdditionalProperties[ddTimestamp] = lr.Timestamp().AsTime().Format(time.RFC3339)
+	}
+
+	var tags = attributes.TagsFromAttributes(res.Attributes())
+	if len(tags) > 0 {
+		tagStr := strings.Join(tags, ",")
+		l.Ddtags = datadog.PtrString(tagStr)
+	}
+	return l
+}
```

Comment:
How are we handling the log message/body? What about arbitrary additional attributes on the OTel log? I would add test coverage for these as well.

Reply:
Message body as is , is not useful. We would expect to parse the body using  [json_parser](https://github.com/observIQ/stanza/blob/main/docs/operators/json_parser.md) or [regex_parser](https://github.com/observIQ/stanza/blob/main/docs/operators/regex_parser.md)) . The parsers convert the body into `attributes` by default . 

All the log attributes are currently added as additional parameters

Reply:
Why is the body not useful? The spec [says](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md#field-body) it "Can be for example a human-readable string message (including multi-line) describing the event in a free form". I would expect in that case at least we would add it

Reply:
Yes , it has full text . But without parsing it , it is blob that isn't useful . I wanted to force clients  to think  about parsing it , rather than it being after thought.     There are two ways to parse the text either use [stanza](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/pkg/stanza) or [datadog](https://docs.datadoghq.com/logs/log_configuration/parsing/?tab=matchers) .  

There are some basic parsing like `severity` or `trace_id` using stanza we should definitely do  before the log is sent to Datadog. There is more true for trace_id , the conversion to 64 bit currently resides in exporter.

Reply:
My initial intent to only support `json logs` , Datadog has extensive documentation on how to json logs . 
Few links for same 
1. https://docs.datadoghq.com/logs/log_collection/go/ 
2. https://docs.datadoghq.com/logs/log_collection/java/?tab=log4j

Reply:
The only processing/parsing the Datadog exporter should do is the minimum required to map payloads to Datadog's logs data model; anything else is out of scope and should be done elsewhere in the data pipeline (e.g. in a receiver or processor). We should nudge people to correctly set up the rest of the pipeline, but it's not the responsibility of the Datadog exporter to do it: an exporter should only handle the mapping and sending.

The OTLP logs `Body` field seems to closely match the Datadog's logs data model `message` field, so I think we should map one to the other. @nhinsch has spent more time than me looking at this so I will defer to him on this.

Reply:
The parsing of json file should happen as part of [filelogreceiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/filelogreceiver) .  Currently `stanza` reads the line into Body. 

The config i was using was (that we would add as recommended way )

```
      filelog:
        include:
            - /var/log/pods/**/*.log
        resource:
          deployment.environment: otel
        operators:
            - id: parse_body
              type: json_parser
              parse_to: body
            - id: parse_log
              type: json_parser
              parse_from: body.log
              severity:
                parse_from: body.severity
            - id: trace 
              type: trace_parser
              trace_id:
                parse_from: body.trace_id
              span_id:
                parse_from: body.span_id
```

Reply:
I think we should support raw logs as well, not just JSON logs, unless there is some compelling reason why we can't. Also, even for JSON logs, the `message` attribute is very important in Datadog. It is the attribute that appears in the UI as the body of the log, and is indexed for search. So, even if the customer is doing client-side parsing of the body into attributes before sending (which I don't think we should assume all customers will do) we would still want the customer to send a message attribute, and from my read of the OTel specification that would be the body.

For the Kubernetes log example I think we would want to preserve the original blob as the message, and if any other attributes are parsed out from that they would be in addition to the message, but wouldn't replace it.

Reply:
There are lots of limitations of `stanza` parser , it can't have different logic based on file format. 

in the above example `message` would be become the additional property , that datadog would use as the body of log.  Ultimately it is the json format that is sent to datadog needs to have either `message` or `msg`, which would be the case here. 

I can add a config option whether to include body or not .

Reply:
Without doing `json` formatted logs , it would be hard for customers to get logs out of k8s, I can demo it quickly

Reply:
I definitely see why a customer would want to use JSON formatted logs, however, my understanding is as follows:

- By the time the log arrives at this function it is a `LogRecord` and we want to convert it to a `datadogV2.HTTPLogItem`
- The `LogRecord` has a `body`
- The `body` should not be discarded, it should be carried over to the Datadog format
- The corresponding field in the Datadog format is the `message` attribute, so we should always convert the `body` to the `message`
- Any other pre-processing the customer does or does not do beforehand in order to make sure the `body` of the `LogRecord` is correct is not the concern of this function, this function only cares about converting a `logRecord` to a `datadogV2.HTTPLogItem`

Does that all make sense to you as well or am I missing something here?

Reply:
See https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/13987#discussion_r973345449 for the outcome of this discussion